Refactor algorithm.rs and switch to the GraphAlgorithm trait

adinapoli-mndc opened the pull request:
This patch switches the `algorithm.rs` module to use the `GraphAlgorithm` trait from `oscoin-graph-api`.

MeBrei commented on src/algorithm.rs:
```diff
@@ -220,20 +199,98 @@ pub fn rank_network<'a, L, G: 'a>(
     Ok(())
 }
 
+/// Marker type used to implement
```
Could you maybe also explain `Mock` vs `Phantom` in your refactoring.

adinapoli-mndc replied:
Yeah, all fair, the documentation part of this PR was a bit neglected in the excitement of getting it out of the door 😉

adinapoli-mndc commented on src/algorithm.rs:
```diff
@@ -344,44 +391,33 @@ mod tests {
+                "id: p3 osrank: 0.2125",
+                "id: a1 osrank: 0.0575",
+                "id: a2 osrank: 0.2675",
+                "id: a3 osrank: 0.065",
```
This is because now we are not _recreating_ the RNG every time (remember we had this discussion already?) but rather we are using the same _initial seed_ for the whole algorithm, this is why (I believe) the ranks now change.

adinapoli-mndc commented on src/types/network.rs:
```diff
@@ -439,6 +449,31 @@ where
+        node_id: &Id<Self::Node>,
+        dir: Direction,
+    ) -> EdgeRefs<Id<Self::Node>, Id<Self::Edge>> {
+        let mut neighbors = Vec::default();
```
Not only that, but I have realised this function was completely bonkers 😬 . I did rewrite it and added proper tests, this time 😉

adinapoli-mndc commented:
@MeBrei I think I have addressed all the comments now. As far as the `osrank` values changed goes, I am inclined to merge this one as it's a big refactoring, and perhaps we dig deeper if we think there is some more cunning bug lurking around, but I think the explanation I have given in the in-line comments is plausible.
